### PR TITLE
Mount propagation in kubelet

### PR DESCRIPTION
**Pull request by jsafrane**
Together with #45724 it implements mount propagation as proposed in https://github.com/kubernetes/community/pull/659

There is:

- New alpha annotation that allows user to explicitly set propagation mode for each `VolumeMount` in pod containers (to be replaced with real `VolumeMount.Propagation` field during beta) + validation + tests. "Private" is the default one (= no change to existing pods).

  I know about proposal for real API fields for alpha feature in https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/edit, but it seems it's not implemented yet. It would save me quite lot of code and ugly annotation.

- Updated CRI API to transport chosen propagation to Docker.

- New `kubelet --experimental-mount-propagation` option to enable the previous bullet without modifying types.go (worked around with changing `KubeletDeps`... not nice, but it's better than adding a parameter to `NewMainKubelet` and removing it in the next release...)

```release-note
kubelet has alpha support for mount propagation. It is disabled by default and it is there for testing only. This feature may be redesigned or even removed in a future release.
```

@derekwaynecarr @dchen1107 @kubernetes/sig-node-pr-reviews

### CODE REVIEW

**Comment by jsafrane** [comment]
And I promise I'll write an e2e or node conformance test when #45724 lands and if there is a way how to add `--test-mount-propagation` to kubelet under test. Any suggestion/precedent?

**Comment by derekwaynecarr** [comment]
@jsafrane -- if you wanted to use this in node e2e test, you can control how kubelet is launched here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/services/kubelet.go#L130

File: pkg/kubelet/apis/cri/v1alpha1/api.pb.go
Comment (derekwaynecarr) [inline pkg/kubelet/apis/cri/v1alpha1/api.pb.go]: cc @yujuhong as primary reviewer for CRI changes.

File: pkg/kubelet/dockershim/helpers.go
Comment (derekwaynecarr) [inline pkg/kubelet/dockershim/helpers.go]: it would be nice if we had a unit test for this function.

File: pkg/kubelet/kubelet.go
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet.go]: i thought this would be a pointer.

File: pkg/kubelet/kubelet_pods.go
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: i am wondering if the kubelet should reject pods that use this annotation during admission if the feature is not enabled.

i know we said we would deprecate the annotation, but i dont always trust our future self.

Comment (jsafrane) [inline pkg/kubelet/kubelet_pods.go]: Doesn't admission controller run on API server? Or kubelet has it's own?

File: pkg/kubelet/dockershim/helpers.go
Comment (jsafrane) [inline pkg/kubelet/dockershim/helpers.go]: unit test added

File: pkg/kubelet/kubelet.go
Comment (jsafrane) [inline pkg/kubelet/kubelet.go]: pointer added

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [inline pkg/kubelet/kubelet_pods.go]: `mountPropagationAdmissionHandler` added

**Comment by jsafrane** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by Random-Liu** [comment]
Reassigning to @dchen1107, because she should have better idea on this.

**Comment by gurvindersingh** [comment]
@jsafrane I am not sure as which one will get merge to master for supporting mount propagation. But I am wondering in this PR case the POD author has to add this annotation himself/herself. Wondering if there is a way to auto annotate PODs or it would be an idea to pick the annotation from PVC like here done for GID https://github.com/kubernetes/kubernetes/blob/1e1040b524dda1c97dad4faab1910972b71a5c79/pkg/volume/util/volumehelper/volumehelper.go#L42 as then if there is an annotation on PVC for propagation mode then PODs using those PVC will use that, POD can override the mode if they prefer by adding annotation themselves if that's preferred.

I would like to have a way where I, as a cluster admin, can control the propagation mode for certain PVC which requires mount propagation e.g. backed by distributed file systems.

**Comment by yujuhong** [comment]
@jsafrane @derekwaynecarr maybe I missed some discussions on this topic, but I think the change to the proposal (https://github.com/kubernetes/community/pull/659) should get reviewed first before we jump into the reviewing the code.

Is this slated for 1.7?

**Comment by jsafrane** [comment]
@yujuhong we have mount containers on sig-storage 1.7 list. The agreed proposal won't work (and it was not implemented anyway), I'm trying to catch 1.7 with my update in https://github.com/kubernetes/community/pull/659

**Comment by jsafrane** [comment]
@gurvindersingh, only HostPath volumes can be slave or shared and the mount propagation is property of the *destination* directory where the HostPath volume is mounted, not the source volume (i.e. PV/PVC). So it does not make much sense to add propagation to PVs/PVCs.

Do you have any specific use case in mind?

**Comment by gurvindersingh** [comment]
@jsafrane PV in this case I mentioned above is hostpath and PVC is the for the same PV. As for security reason we don't allow users to use hostPath directly but a cluster admin creates PV for hostpath and then users can use them using PVC (which are backed by hostPath PV). In my use case I have distributed file system like "GlusterFS" mounted on all worker nodes and then have hostpath PV created from certain directory path depending upon the project. On those PVs, I would like to have "rslave" mode attached as when POD using them with PVCs can get mount/remount events propagated in the case of issue with storage/network layer. So wondering if it make sense to add the mount propagation annotation on PV/PVC to be automated by cluster admin on certain PVs/PVCs.

**Comment by jsafrane** [comment]
rebased and fixed verification test

**Comment by jsafrane** [comment]
@gurvindersingh, sorry for late reply. Initially I thought that we should start with "private" as default and then move to "rslave". Now I see a real use case for "rslave" as default, so let's try to default to "rslave" from start and fail early. "rslave" as default was already envisioned in original [propagation.md](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/propagation.md)

I updated https://github.com/kubernetes/community/pull/659, I'll update (and rebase) this PR shortly.

**Comment by jsafrane** [comment]
- rebased
- made rslave default when --test-mount-propagation is set (otherwise defaults to private as in previous Kubernetes releases)

File: cmd/kubelet/app/options/options.go
Comment (vishh) [inline cmd/kubelet/app/options/options.go]: @mtaufen what's the new policy for experimental flags?

File: cmd/kubelet/app/options/options.go
Comment (ddysher) [inline cmd/kubelet/app/options/options.go]: should we follow other flag convention and use `experimental` instead?

File: pkg/api/helper/helpers.go
Comment (ddysher) [inline pkg/api/helper/helpers.go]: Add godoc for exported function?  I think some lint tools will complain about this...

File: cmd/kubelet/app/options/options.go
Comment (jsafrane) [inline cmd/kubelet/app/options/options.go]: It has been decided on sig-node meeting that `--test` is the right prefix here.

File: cmd/kubelet/app/options/options.go
Comment (mtaufen) [inline cmd/kubelet/app/options/options.go]: We talked about this specific flag in sig-node and agreed it was ok for now. Can you please name it `ExperimentalMountPropagation` though (and the flag `--experimental-mount-propagation`), just to make it really clear that it's experimental?

We don't want to keep adding experimental stuff to the Kubelet's versioned configuration types at the moment, so the only option is to add flags for experimental things. At some point (hopefully soon) apimachinery will have field gates (https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/edit#), and we can probably use that to include experimental things in the versioned configuration type (without having to rev the version), rather than as flags.

Comment (jsafrane) [inline cmd/kubelet/app/options/options.go]: renamed to `--experimental-mount-propagation` (we agreed on `--test-*` on the meeting though...)

File: pkg/api/helper/helpers.go
Comment (jsafrane) [inline pkg/api/helper/helpers.go]: good catch, fixed.

File: cmd/kubelet/app/options/options.go
Comment (jsafrane) [inline cmd/kubelet/app/options/options.go]: and now the decision changed, renamed to `--experimental-mount-propagation`

**Comment by jsafrane** [comment]
- fixed review remarks
- reworked the e2e test to run on Docker >= 1.13. Feel free to suggest a better way. Am I correct that node e2e tests always run with Docker and never with rkt?

As consequence, the test is skipped on all images that run node e2e tests. Is there a plan to move to cos-60-beta, where the test could run?

File: cmd/kubelet/app/options/options.go
Comment (ddysher) [inline cmd/kubelet/app/options/options.go]: thanks, this is clearer IMO.

**Comment by kfox1111** [comment]
#46643 raises an interesting question. Maybe hostPath isn't the only volume plugin that should support shared mounts. Its also useful for emptyDir. Think fuse file system in a sidecar and shared via emptyDir to the rest of its pod.

**Comment by croomes** [comment]
This works well with a StorageOS DaemonSet (https://gist.github.com/croomes/2019ed1f4178bd199bf7c65e6379dcaa).  Would be happy to help with other testing...

File: pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto
Comment (feiskyer) [inline pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto]: nit: add `.` at end.

**Comment by luxas** [comment]
ping @kubernetes/sig-node-pr-reviews 

@mtaufen Why aren't we using a feature gate for this?

**Comment by jsafrane** [comment]
Updated according to changes in https://github.com/kubernetes/community/pull/659

- Alpha feature flag is used instead of `--experimental-mount-propagation`
- `MountPropagation` field moved from `struct VolumeMount` (part of Pod definition) to `HostPathVolumeSource` (part of PersistentVolume or Pod)
- No alpha annotations

(The last commit with e2e_node test still needs some changes, will fix next week)

**Comment by jsafrane** [comment]
I dropped the e2e test. The test needs MountPropagation=true feature gate but node e2e test runs internal API server without any FeatureGates. I will post a separate PR that will fix that and add MountPropagation test when this one is merged.

**Comment by jsafrane** [comment]
Updated according to latest version of https://github.com/kubernetes/community/pull/589

File: pkg/apis/settings/validation/validation.go
Comment (saad-ali) [inline pkg/apis/settings/validation/validation.go]: nit: put a comment e.g. `/* paramName */` next to nil so it is clear what is being set to `nil`.

Comment (saad-ali) [inline pkg/apis/settings/validation/validation.go]: nit: For clarity, add a comment, e.g. `nil /* paramName */`, to indicate what parameter `nil` is being passed in for.

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

This looks good from storage-sig side. Thanks for all your work @jsafrane!

@kubernetes/sig-node-pr-reviews can we please get a review ASAP, we want to get this in and let it bake ASAP.

**Comment by derekwaynecarr** [comment]
@dchen1107 @vishh -- did you guys have cycles to do the kubelet review for this, if not I will pick up.

File: api/openapi-spec/swagger.json
Comment (vishh) [inline api/openapi-spec/swagger.json]: nit: "mountPropagation determines how mounts are propagated from the host to container and the other way around"

File: pkg/api/types.go
Comment (vishh) [inline pkg/api/types.go]: nit: nit: "mountPropagation determines how mounts are propagated from the host to container and the other way around. When not set, defaults to MountPropagationHostToContainer."

File: pkg/features/kube_features.go
Comment (vishh) [inline pkg/features/kube_features.go]: Is it really tied to just HostPath volumes?

File: pkg/kubelet/kubelet_pods.go
Comment (vishh) [inline pkg/kubelet/kubelet_pods.go]: nit: Can this logic be in a separate method?

**Review by vishh — CHANGES_REQUESTED** [review]
Just a few comments and nits. This is mostly ready to be merged!

File: pkg/api/types.go
Comment (jsafrane) [reply pkg/api/types.go]: @bgrant0607 was against using world "default" when something does not fill the default in `default.go`. https://github.com/kubernetes/kubernetes/pull/19921#issuecomment-176575106

File: api/openapi-spec/swagger.json
Comment (jsafrane) [reply api/openapi-spec/swagger.json]: fixed

File: pkg/features/kube_features.go
Comment (jsafrane) [reply pkg/features/kube_features.go]: fixed

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: extracted into a separate method

**Comment by jsafrane** [comment]
@vishh, thanks for your review. I fixed most of your comments, PTAL

File: pkg/api/types.go
Comment (bgrant0607) [reply pkg/api/types.go]: Correct. Is there a reason we wouldn't fill in the default?

File: api/openapi-spec/swagger.json
Comment (vishh) [reply api/openapi-spec/swagger.json]: It looks like the text hasn't been updated yet?

**Comment by vishh** [comment]
Only a single [nit](https://github.com/kubernetes/kubernetes/pull/46444/files#r135321885) is unresolved. This is good to go otherwise. I'll approve it as is.
@saad-ali and others wait till all the open comments are resolving before marking this pr with lgtm label.

/approve

File: api/openapi-spec/swagger.json
Comment (jsafrane) [reply api/openapi-spec/swagger.json]: It is now, sorry.

**Comment by jsafrane** [comment]
I got approval from sig-node and sig-storage, now the API
/assign @thockin @bgrant0607

**Comment by jsafrane** [comment]
(I'll rebase when I get the approvals, it conflicts too fast)

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: Should this be Forbidden ?

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: Why passing `mountPropagationEnabled` multiple times when instead, you could check that directly in `translateMountPropagation()`?

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: s/Invalid/invalid
Error should start with a lower-case letter

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: remove the extra line.

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: remove the extra line.

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: remove the extra line.

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: use `format.Pod(pod)` to log pod-level information.

Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: done

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: done

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: done

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: done

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: fixed

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: Good question. Initially there was `kubelet --experimental-mount-propagation` that needed passing all over the place and this is the last remainder. Removed.

**Comment by jsafrane** [comment]
@yujuhong fixed your comments and rebased, PTAL

File: pkg/api/validation/validation.go
Comment (jsafrane) [reply pkg/api/validation/validation.go]: fixed

File: staging/src/k8s.io/api/core/v1/types.go
Comment (yujuhong) [inline staging/src/k8s.io/api/core/v1/types.go]: s/Optional/+optional

File: pkg/api/validation/validation.go
Comment (yujuhong) [inline pkg/api/validation/validation.go]: The api changes guideline for alpha fields also suggestss clearing the field before persisting to the storage. https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version

Not sure if it's needed but it's worth checking out.

File: pkg/api/validation/validation.go
Comment (yujuhong) [inline pkg/api/validation/validation.go]: The convention seems to be using `Forbidden`

File: pkg/kubelet/dockershim/helpers_test.go
Comment (yujuhong) [inline pkg/kubelet/dockershim/helpers_test.go]: The test file uses assert/require. It'd be good to be consistent.
`assert.Equal(t, result, expectedResult, result)`

**Review by vishh — APPROVED** [review]

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: If the feature is enabled, there is no way to specify `PRIVATE` and the default changes to HOST_TO_CONTAINER. 

This *might* be fine for Alpha, but can we actually change the default mode in Beta/GA? Sorry I didn't follow the PR proposal closely, so most likely this has been answered before.

**Comment by yujuhong** [comment]
Have a question and some nits. Holding the lgtm for now.

File: staging/src/k8s.io/api/core/v1/types.go
Comment (jsafrane) [reply staging/src/k8s.io/api/core/v1/types.go]: ok

File: pkg/api/validation/validation.go
Comment (jsafrane) [reply pkg/api/validation/validation.go]: ok

File: pkg/kubelet/dockershim/helpers_test.go
Comment (jsafrane) [reply pkg/kubelet/dockershim/helpers_test.go]: ok

File: pkg/kubelet/kubelet_pods.go
Comment (jsafrane) [reply pkg/kubelet/kubelet_pods.go]: The proposal expects that HOST_TO_CONTAINER is the new default and PRIVATE is not needed at all. We need to test this theory in alpha.

File: pkg/api/validation/validation.go
Comment (jsafrane) [reply pkg/api/validation/validation.go]: I did not find any feature gate checks in `pkg/registry/core`. Is really required?

Comment (jsafrane) [reply pkg/api/validation/validation.go]: there is discussion about clearing alpha fields in https://github.com/kubernetes/kubernetes/pull/50924. I'll wait for result and fix mountPropagation field when necessary.

**Comment by jsafrane** [comment]
@yujuhong fixed all your comments but clearing alpha fields, PTAL. Thanks!

File: pkg/kubelet/kubelet_pods.go
Comment (yujuhong) [reply pkg/kubelet/kubelet_pods.go]: I'm slightly concerned that we won't gain enough confidence through feedback of this alpha feature. Changing the default behavior is always tricky.
I don't want to block the PR since the proposal has been through rounds of reviews. We can come back to this before graduating the feature to beta.

File: pkg/api/validation/validation.go
Comment (hzxuzhonghu) [inline pkg/api/validation/validation.go]: why no err returned about nil

Comment (jsafrane) [reply pkg/api/validation/validation.go]: See the comment two lines below. Some `PodPreset` test was failing on it as it validated a PodSpec with no containers.

Comment (jsafrane) [reply pkg/api/validation/validation.go]: ... some PodPreset test was failing on it as it validated **VolumeMounts** with no containers.

**Comment by jsafrane** [comment]
restoring lgtm after rebase of generated stuff and features.go

**Comment by jsafrane** [comment]
rebased, restoring lgtm again

File: pkg/api/validation/validation.go
Comment (hzxuzhonghu) [reply pkg/api/validation/validation.go]: If container is previously  validated,  I do not think this `if container == nil `  make  sense

**Pull request merged**
